src/photon/photon_algorithm.c

```diff
@@ -355,22 +412,64 @@ void handle_object_available(local_scheduler_state *state,
     CHECK(event_loop_remove_timer(state->loop, fetch_req->timer) == AE_OK);
     free(fetch_req);
   }
+
+  /* Move any tasks whose object dependencies are now ready to the dispatch
+   * queue. */
+  /* TODO(swang): This can be optimized by keeping a lookup table from object
+   * ID to list of dependent tasks in the waiting queue. */
+  task_queue_entry *elt, *tmp;
+  DL_FOREACH_SAFE(algorithm_state->waiting_task_queue, elt, tmp) {
+    if (can_run(algorithm_state, elt->spec)) {
+      LOG_DEBUG("Moved task to dispatch queue");
+      DL_DELETE(algorithm_state->waiting_task_queue, elt);
+      DL_APPEND(algorithm_state->dispatch_task_queue, elt);
+    }
+  }
+
+  /* Try to dispatch tasks, since we may have added some from the waiting
+   * queue. */
+  dispatch_tasks(state, algorithm_state);
 }
 
-void handle_object_removed(local_scheduler_state *state, object_id object_id) {
+void handle_object_removed(local_scheduler_state *state,
+                           object_id removed_object_id) {
   scheduling_algorithm_state *algorithm_state = state->algorithm_state;
   available_object *entry;
-  HASH_FIND(handle, algorithm_state->local_objects, &object_id,
-            sizeof(object_id), entry);
+  HASH_FIND(handle, algorithm_state->local_objects, &removed_object_id,
+            sizeof(removed_object_id), entry);
   if (entry != NULL) {
     HASH_DELETE(handle, algorithm_state->local_objects, entry);
     free(entry);
   }
+
+  /* Move dependent tasks from the dispatch queue back to the waiting queue. */
```

Comment:
It can happen if an object is evicted. This commit includes a test case in `photon_tests.c` that tests this scenario.